keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,1128 @@
+}
+```
+
+Skipping multiple stack levels at once via `WithCallDepth` is not working with
+loggers that output via `testing.T.Log`. `WithCallDepth` therefore should only
+be used by code in `test/e2e` where it can be assumed that the logger is not
+using `testing.T`.
+
+
+### [logcheck](https://github.com/kubernetes/klog/tree/main/hack/tools/logcheck)
+
+That tool is a linter for log calls. It’s used to check that log calls are
+well-formed. Because it is used in core Kubernetes and SIGs (for example,
+[metrics-server](https://github.com/kubernetes-sigs/metrics-server/blob/4b20c2d43e338d5df7fb530dc960e5d0753f7ab1/Makefile#L252-L257),
+it should better be hosted in a repo where "go install" is fast. It only
+depends on the standard Go runtime, therefore it can be moved to
+`k8s.io/klogr/logcheck`. The tool will be updated to:
+- detect usage of klog in code that should have been converted to klogr,
+- check not only klog calls, but also calls through the `logr.Logger` interface,
+- detect direct calls to `WithValue`, `WithName`, and `NewContext` where
+  the `klogr` wrapper functions should be used instead,
+- detect function signatures with both `context.Context` and `logr.Logger`.
+
+#### Code examples
+
+See https://github.com/pohly/kubernetes/compare/master-2022-01-12...pohly:log-contextual-2022-01-12
+for the helper code and a tentative conversion of some parts of client-go and
+kube-scheduler. The last commit removes the klog dependency from `k8s.klogr`
+and thus demonstrates that the final goal of this KEP can be achieved.
+
+##### klogr replacing klog completely, explicit logger parameter
+
+```diff
+diff --git a/pkg/scheduler/framework/plugins/volumebinding/binder.go b/pkg/scheduler/framework/plugins/volumebinding/binder.go
+index ee4989be0b7..3ae0d0b4aaa 100644
+--- a/pkg/scheduler/framework/plugins/volumebinding/binder.go
++++ b/pkg/scheduler/framework/plugins/volumebinding/binder.go
+@@ -44,7 +44,7 @@ import (
+        storagehelpers "k8s.io/component-helpers/storage/volume"
+        csitrans "k8s.io/csi-translation-lib"
+        csiplugins "k8s.io/csi-translation-lib/plugins"
+-       "k8s.io/klog/v2"
++       "k8s.io/klogr"
+        v1helper "k8s.io/kubernetes/pkg/apis/core/v1/helper"
+        pvutil "k8s.io/kubernetes/pkg/controller/volume/persistentvolume/util"
+        "k8s.io/kubernetes/pkg/features"
+@@ -151,7 +151,7 @@ type InTreeToCSITranslator interface {
+ type SchedulerVolumeBinder interface {
+        // GetPodVolumes returns a pod's PVCs separated into bound, unbound with delayed binding (including provisioning)
+        // and unbound with immediate binding (including prebound)
+-       GetPodVolumes(pod *v1.Pod) (boundClaims, unboundClaimsDelayBinding, unboundClaimsImmediate []*v1.PersistentVolumeClaim, err error)
++       GetPodVolumes(logger klogr.Logger, pod *v1.Pod) (boundClaims, unboundClaimsDelayBinding, unboundClaimsImmediate []*v1.PersistentVolumeClaim, err error)
```

Comment:
I'd argue to put logger and things at the end of the list, personally, but we can defer that discussion until PRs